Implement Postgres LISTEN interface.

thedodd opened the pull request:
This changeset introduces an interface for using PostgreSQL's LISTEN
functionality from within sqlx.

The listen interface is implemented over the PgConnection, PgPool & the
PgPoolConnection types for ease of use. In the case of PgPool, a new
connection is acquired, and is then used for LISTEN functionality.

Closes #87

#### todo
- [x] `PgListener` conn is dedicated to receiving notifications only, so ensure it is properly encapsulated.
- [x] No notification buffering, though users could implement buffering on their own if needed.
- [x] Use `Ext` traits for everything.
- [x] Expose only a `.listen(&[&str])` interface on `PgConnection`, `PgPoolConnection` & `PgPool` (covers single & multi channel).
- [x] Isolate all of the business logic of the `PgListener` in a single `receive` method.
- [x] Impl reconnect behavior when created from a `PgPool`.
- [x] Impl `Stream` in terms of `receive`. 
- [x] docs. Added an example app.

thedodd commented:
Haven't written any tests for this yet. Happy to do so, but I will have to tackle that over the weekend probably.

thedodd commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,188 @@
+                Some(Message::NotificationResponse(notification)) => return Ok(notification.into()),
+                // TODO: verify with team if this is correct. Looks like the connection being closed will cause an error
+                // to propagate up from `recevie`, but it would be good to verify with team.
+                Some(_) | None => continue,
```
@mehcode && @abonander just wanted to verify this with you guys. Thoughts?

abonander replied:
`None` means the connection returned `ReadyForQuery`, so it really just means there's no more messages since the last request. That's probably the right call here but I also want to say that any unexpected messages _should_ be a logic error and maybe shouldn't just get eaten, especially if we're still allowing use of the connection.

thedodd replied:
@abonander do you think I should just raise a protocol error here & then close the connection if we receive something other than a None or a NotificationResponse here?

thedodd replied:
Raising a protocol error, will push the changeset shortly for additional review.

abonander commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,188 @@
+    pub async fn listen(&self, channel: &impl AsRef<str>) -> Result<PgListener<PgPoolConnection>> {
+        let mut conn = self.acquire().await?;
+        let cmd = PgConnection::build_listen_query(channel);
+        let _ = conn.execute(cmd.as_str(), Default::default()).await?;
```
`.execute()` isn't the right call here because `LISTEN` isn't a preparable statement. Use `.send()` instead (which also doesn't require constructing a `PgArguments`).

abonander replied:
(Same thing with all the other usages of `.execute()`.)

thedodd replied:
Sounds good.

abonander commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,188 @@
+        channels: impl IntoIterator<Item = impl AsRef<str>>,
+    ) -> Result<PgListener<Self>> {
+        for channel in channels {
+            let cmd = format!(r#"LISTEN "{}""#, channel.as_ref());
```
It'd be neat to buffer all these into one string, delimited by semicolons.

thedodd replied:
Will do.

abonander commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    /// Build a LISTEN query based on the given channel input.
+    fn build_listen_query(channel: &impl AsRef<str>) -> String {
```
This can just be a free function in the module. I'd do a version for `listen_all()` as well to dedup some code.

thedodd commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,188 @@
+        Ok(PgListener::new(self))
+    }
+
+    /// Fetch a new connection from the pool and register it as a listener on the specified channels.
```
I just noticed that I need to update these comments. Will take care of that here soon.

abonander commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,188 @@
+    }
+}
+
+impl<C> std::ops::DerefMut for PgListener<C> {
```
If we're going to allow use of `PgConnection` with active subscriptions, we need to patch either `.receive()` or usages of it to push `NotificationResponse` messages to a `Vec` because they might be interleaved with responses from other commands. And then `PgListener.next()` should check that vec first before calling `.receive()`.

thedodd replied:
Yea, I was think along those lines as well. On one hand, if the connection is in active use, then perhaps we don't worry about them, and we only handle notifications received while actually waiting for notifications.

On the other hand, we could create a buffer of the events as you mentioned. I would be hesitant to create an unlimited buffer of events because that could bloat the vector pretty heavily depending on the use case. I definitely see async notifications as being a bit transient as well. Some users may consider buffering of the events to be an anti-pattern, where they only want to see events as they occur. So, leaving it up to the user may be best.

What do you think about creating a configuration parameter in the `listen/listen_all` methods which will allow the user to configure message buffering? Perhaps an enum like:

```rust
enum BufferStrategy {
    All,
    Max(u32),
    None,
}
```

Then we just update the `PgConnection` state to embed this as an optional configuration (only `Some(_)` if `listen/listen_all` was used on the connection) — say, `buffer_strategy: Option<BufferStrategy>` — and then we buffer using the given strategy.

The only real complication with this strategy is that we will have to update the signature of `PgConnection.receive` to take some sort of parameter — say, `is_listening: bool` (or whatever) — so that it knows not to buffer when it is being called from the `PgListen.next` method.

The algorithm for that subsection of receive would be:
- if a NotificationMessage is received on the wire and `buffer_strategy.is_none()` then drop the notification.
- if a NotificationMessage is received on the wire and `buffer_strategy.is_some()` and `is_listening`, then immediately return the notification.
- if a NotificationMessage is received on the wire and `buffer_strategy.is_some()` and `!is_listening`, then buffer the notification using the configured strategy.

With that, we would just need to update `PgListener.next` to consult the notifications buffer before attempting to call `receive`. That may work pretty well. Leaving it up to the user is always nice too.

@abonander @mehcode thoughts?

abonander replied:
I'm leaning toward agreeing with this approach, although what about a `.buffer_limit(limit: impl Into<Option<usize>>)` method on `PgListener` instead? `None` would be "don't buffer" and the user would have to explicitly choose to make the buffer (effectively) unbounded with `usize::max_value()`.

Are we also thinking the buffer should be a FIFO queue with the oldest messages being popped first when the limit is reached? That way the user can set a limit of 1 and effectively have a "latest message only" subscription, although for that case I think supporting "latest message per topic" is strictly more useful (but can be implemented later).

abonander replied:
@mehcode looked through `libpq` to see what they do, and it looks like they just push messages to an unbounded linked list: https://github.com/postgres/postgres/blob/7559d8ebfa11d98728e816f6b655582ce41150f3/src/interfaces/libpq/fe-protocol3.c#L1415

So we can already do them one better by using a more datastructure for real computer architecture (`Vec`/`VecDeque`).

thedodd replied:
Yea, double ended queue is definitely the way to go here, drop items off of the front as newer items come in which would breach the capacity.

Adding the `.buffer_limit(..)` method is a good idea. API would be something like:
- if not specified, the default will be an unbounded deque.
- if `None` is provided, then buffering will be disabled.
- if `Some(usize)` is provided, then we limit the capacity of the buffer to the specified size.

Sound about right? As far as the logging ... maybe we just expose a method so that the user can check the buffer size themselves, and if they want to log that information, we allow them to do so. Something like `PgListener.buffer_size() -> Option<usize>` (option because buffering may be disabled). Thoughts?

thedodd replied:
Ok, this is effectively resolved now per: https://github.com/launchbadge/sqlx/pull/98#issuecomment-580370236

abonander commented:
Re: tests, yes please, whenever it's convenient.

thedodd commented:
Ok, addressed all of the review comments and also put together a quick and dirty demo on my local machine to ensure the API was working as needed. Had to update the `AsMut<_>` bit to `DerefMut<_>` as that is already implemented by all potential types of the PgListener's `C` type param.

Only outstanding item at this point is the discussion here: https://github.com/launchbadge/sqlx/pull/98#discussion_r372688740 about buffering of notifications. Should be quite simple to implement, I just want to get some feedback from you guys before moving forward on one of the patterns. 

Cheers.

mehcode commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,216 @@
+fn build_listen_all_query(channels: impl IntoIterator<Item = impl AsRef<str>>) -> String {
+    channels.into_iter().fold(String::new(), |mut acc, chan| {
+        acc.push_str(r#"LISTEN ""#);
+        acc.push_str(chan.as_ref());
```
Channels must be a quoted identifier (good) but also must be no more than 63 characters. Should we error here? Postgres just truncates :crying_cat_face:

thedodd replied:
I can't seem to find that in the postgres docs anywhere. Does NOTIFY also truncate the channel? If so, perhaps we just include this information in a doc string and trust the user to do the correct thing? 

Maybe we just do a check and log a warning on the standard logging facade?

mehcode commented:
For context on buffering and API design:

 * **libpq** (official reference impl, C) uses an internal linked that grows without bound

 * **psycopg2** (Python) uses a list that it appends without bound and you receive messages asynchronously by polling

 * **go/pg** (Go) prevents you from using the connection apart from receiving messages and has you receive messages inline (as we are in this PR)

 * **vertx-sql-client** (Java) allows you to use the connection and sends messages to preregistered callback functions

 * **node-postgres** (Node) has you register an event handler for `notification`s. Its' not clear how you poll for new ones.

 * **JDBC** (more Java) allows you to poll a `getNotifications()` that returns a list of notifications since the last time you called it

---

My recommendation would be no buffering and prevent external usage of the connection (e.g., remove the derefs). An application could make use of their shared pool to checkout another connection to do work on a notification.

My gut says this is the recommended approach when considering high throughput. You don't want to lock up the listening connection by using it for something else. As a a parallel, you don't want to lock up the `TcpListener` and should immediately spawn with the newly created `TcpStream` object.

---

Here is a theoretical example:

```rust
let mut listener: PgListener = pool.listen("channel").await?;

while let Some(message) = listener.receive().await? {
  // Immediately spawn so we can receive the next message
  task::spawn(async move {
    // Handle message
  })
}
```

---

API design thoughts:

 * Keep database-specific extensions to `Ext` traits. Maybe some other database driver wants to use the name `Pool::listen` for something.

 * Keep all _work_ inside of `PgListener::receive`. E.g., keep `listen` sync and infallible. This lets people use `.listen("foo").listen("other")` as a builder cleanly and additionally we should be collected channels to a vec anyway so we can implement reconnections in the future.

 * Re-connections. Not something we should block on having the feature at all but we should think how this affects the API. I would expect a dedicated `PgListener` type to internally handle re-connections if I make it from a `Pool`. 

##### `PgListener`

```rust
struct PgListener { }

impl PgListener {
  // Listen to _one_ more channel
  fn listen(mut self, channel: &str) -> Self;

  // Wait for a notification
  async fn receive(&mut self) -> sqlx::Result<PgNotification<'_>>;

  // Check if there is an available notification without blocking
  fn try_receive(&mut self) -> Option<PgNotification<'_>>;

  // Close the listener and its owned connection
  async fn close(self) -> sqlx::Result<()>;
}
```

##### `PgNotification`

```rust
#[derive(Debug)]
struct PgNotification<'a> {
  // We should be able internally re-use a work buffer for this payload
  pub payload: &'a str,

  // Postgres process ID that sent this
  // We should probably surface the one in `PgConnection` in a later task
  pub process_id: i32,
}
```

##### `PgConnectionExt`

```rust
#[async_trait]
trait PgConnectionExt {
  fn listen(self, channel: &str) -> PgListener;
}

impl<C> PgConnectionExt for C where C: Connection<Backend = Postgres> { ... }
```

##### `PgPoolExt`

```rust
#[async_trait]
trait PgPoolExt {
  fn listen(&self, channel: &str) -> PgListener;
}

impl PgPoolExt for Pool<Postgres> { ... }
```

mehcode commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,216 @@
+                    )
+                    .into())
+                }
+                None => continue,
```
`None` from `receive` means the connection was closed I believe.

thedodd replied:
I think this reveals an aspect of this design overall which makes me think that we should be treating this as an actual `Stream`. It could still be used in all of the same ways which have been suggested, but semantically this pattern is a stream pattern whether we use a `Stream` implementation or not.

If we use a stream implementation, then returning `None` is the logical ending point of the stream. Without using a stream, users may end up in a state of confusion about how to handle conditions where the underlying connection has been closed. Should this interface just continually return an error if `None` is hit here on `L111`? What if the user doesn't handle it properly?

With a stream impl, users can still do:
```rust
while let Some(notification) = notify_stream.next().await {
// ...
}
```
See https://docs.rs/futures/0.3.1/futures/stream/trait.StreamExt.html#method.next for more details.

We could even design the stream to be `Stream<Item=PgNotification>`; IE, no error condition. If an error comes up within the stream impl, we can just transform it into a `None` (as there should really be no actual error conditions in the listener). If running from a pool, we can retrieve a new connection and resume listening any time a connection is lost.

With the non-stream interface, we are exposing a stream-like interface with none of the standard expectations which folks will have when working with async Rust. The implementation of the stream interface should be pretty simple. I'm thinking we could use [stream::unfold](https://docs.rs/futures/0.3.1/futures/stream/fn.unfold.html) to wrap the calls to `receive`. The init state would just be `()` probably, as we don't really need a seed value. Using unfold just allows us to stay in async/await land instead of having to work with `poll_next` (which can be a PITA).

@mehcode @abonander thoughts?

thedodd replied:
Ok, going to get this all wrapped up in the `PgListener.receive` interface (which the stream impl will be based upon as well). I have a TODO item for it at the top of the PR now as well.

mehcode commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,216 @@
+    }
+}
+
+impl<C: Connection<Database = Postgres>> crate::Executor for PgListener<C> {
```
Not something you should fix in your PR but I'm tired of how much this trait is delegated. We need a macro. Haha.

mehcode commented:
Awesome work so far though @thedodd :tada: It looks pretty good :100: . We just need to tweak some of the details I think.

thedodd commented:
@mehcode I love the auto-reconnect suggestion when created from a Pool. I'll knock that out for sure. Dead simple to do. Keeping all of the work isolated inside of `PgListener::receive` (or whatever we choose to call the method) is solid, I'll refactor that as well.

----

One thing I wanted to discuss:

> My recommendation would be no buffering and prevent external usage of the connection (e.g., remove the derefs). An application could make use of their shared pool to checkout another connection to do work on a notification.

> My gut says this is the recommended approach when considering high throughput. You don't want to lock up the listening connection by using it for something else. As a a parallel, you don't want to lock up the TcpListener and should immediately spawn with the newly created TcpStream object.

I definitely see the argument, but I will say that in my experience using the `go/pg` is that it is quite annoying to have to use a separate connection to do work (that is a subjective opinion on my behalf, to be sure `:)`). Especially if the workload is not concerned about maximizing throughput.

Perhaps a good middle ground on the design may be to leave this up to the user as well. If the user wants to use the listener solely for receiving notifications and spawning other tasks, they can totally do that, and as they will be (should be) using a pool in such cases, that should be no problem. 

We can further improve the interface by making PgListener generic over one additional type param, and then `impl<C, P=PgPool> PgListener<C, P> { pub fn pool(&self) -> PgPool }` so that when a PgListener is created from a PgPool, there will be a nice convenience function to get a handle to that pool. If the listener is not created from a pool, then that interface will not be present. Thoughts?

----

One additional item: sounds like we need to come to a consensus on the buffering story. There are a few different opinions on how this should be handled so far. Based on [the discussion here](https://github.com/launchbadge/sqlx/pull/98#discussion_r373033482), we should be able to implement a pretty effective strategy for buffering while keeping control in the user's hands. I am pretty much always partial to leaving it up to the user as much as is reasonably possible, as that just makes things more flexible for everyone. The interface will still be pretty simple either way. Thoughts?

thedodd commented:
Per some extensive discussion with @mehcode in the discord channel, here is where we landed:

- `PgListener` conn is dedicated to receiving notifications only.
- No notification buffering, though users could implement buffering on their own if needed.
- Expose only a `::listen(&[&str])` interface on `PgConnection`, `PgPoolConnection` & `PgPool` (covers single & multi channel).
- Isolate all of the business logic of the `PgListener` in a single `receive` method.
- Impl `Stream` in terms of `receive`. 
- Impl reconnect behavior when created from a `PgPool`.
- Use `Ext` traits for everything.

thedodd commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,302 @@
+    pub channel: String,
+    /// The payload of the notification.
+    pub payload: String,
+}
```
@mehcode you mentioned some stuff about the structure of this guy in an earlier review. Not sure if this is what you wanted to see. If not, we could just wrap the `Box<NotificationResponse>` and provide accessors.

thedodd commented on Cargo.lock:
```diff
@@ -4,2268 +4,2275 @@
 name = "aho-corasick"
```
I think this is all due to an update of the Cargo.lock format ... I'm happy to revert this change (which was unintentional). Not that it is actually used by consumers of this library.

thedodd commented:
@mehcode && @abonander just wanted to ping you guys to see if you have time for another round of review. I'm happy to start on modifications right away once review is done. I don't want to hold up 0.3 `:)`. Also, I'm happy to chat on the discord channel if you guys want to fast-track any deeper design changes and such.

thedodd commented on sqlx-core/src/postgres/listen.rs:
```diff
@@ -0,0 +1,297 @@
+    pub async fn close(self) -> BoxFuture<'static, Result<()>> {
+        self.connection.close()
+    }
+}
```
Not sure what you all think about this, but we could wrap the connection in an `Option` so that we can take the connection out and spawn a task to call this `close` method from inside of a `Drop` impl. That way user's wouldn't have to worry about forgetting to call `close` on their own.

With the `PgPoolListener` that is easy (already implemented below), as it wraps the current connection in an `Option` for internally required reasons.

Thoughts?

mehcode commented:
@thedodd, Sorry for the radio silence. I'm still working on the 0.3 changes. Hopefully will be opening a PR for it soon. Don't worry about holding up the 0.3 release. We'll get there.

thedodd commented:
@mehcode want me to go ahead and rebase this? I remember you saying something about potentially already having done this while you were testing things out. I'm happy to rebase.

mehcode commented:
@thedodd No worries. I'll open a PR with the rebase plus a few ergonomic changes soon and we can discuss there.

mehcode commented:
This was merged in via #131
